File: ios/PatientAssignment.swift
import Foundation
import OSLog
import Supabase

/// Kinds stored in `patient_assignments.type`. Extra cases are reserved
/// for upcoming diary and Patient Mode work.
enum PatientAssignmentType: String, Codable, Sendable {
    case questionnaire
    case diaryOne = "diary_one"
    case diaryTwo = "diary_two"
}

enum PatientQuestionnaireSubmitError: LocalizedError {
    case alreadyCompleted
    case cancelled
    case accessDenied
    case invalidAnswers
    case failed

    var errorDescription: String? {
        switch self {
        case .alreadyCompleted, .failed, .invalidAnswers:
            L10n.patientQuestionnaireSubmitError
        case .cancelled:
            L10n.patientQuestionnaireCancelledError
        case .accessDenied:
            L10n.patientQuestionnaireAccessDeniedError
        }
    }
}

/// Request body for `submit-patient-questionnaire`. CamelCase only;
/// patient/therapist/session IDs are not sent.
private struct SubmitPatientQuestionnaireRequest: Encodable {
    let assignmentId: UUID
    let gad7Answers: [Int]
    let phq9Answers: [Int]
    let interferenceLevel: Int
}

private struct SubmitPatientQuestionnaireResponse: Decodable, Sendable {
    let success: Bool?
    let combinedMoodId: Int?
    let sessionId: UUID?
}

enum PatientAssignmentError: LocalizedError {
    case notConfigured
    case notSignedIn
    case invalidIdentifier
    case patientNotConnected

    var errorDescription: String? {
        switch self {
        case .notConfigured: AuthError.notConfigured.localizedDescription
        case .notSignedIn: L10n.therapistDisplayNameNotSignedInError
        case .invalidIdentifier: L10n.questionnaireAssignmentSendError
        case .patientNotConnected: L10n.patientNotConnectedTitle
        }
    }
}

/// A row in `public.patient_assignments`. Identifiers are not shown in UI.
nonisolated struct PatientAssignment: Decodable, Sendable {
    let id: UUID
    let patientId: UUID
    let therapistId: UUID?
    let sessionId: UUID?
    /// Raw `type` column. Unknown future values must not fail decoding.
    let typeValue: String
    let createdAt: Date
    let completedAt: Date?
    let cancelledAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case patientId = "patient_id"
        case therapistId = "therapist_id"
        case sessionId = "session_id"
        case typeValue = "type"
        case createdAt = "created_at"
        case completedAt = "completed_at"
        case cancelledAt = "cancelled_at"
    }

    var type: PatientAssignmentType? { PatientAssignmentType(rawValue: typeValue) }

    var isOpen: Bool { completedAt == nil && cancelledAt == nil }
}

private struct NewPatientAssignment: Encodable {
    let patientId: UUID
    let therapistId: UUID
    let sessionId: UUID
    let type: PatientAssignmentType

    enum CodingKeys: String, CodingKey {
        case patientId = "patient_id"
        case therapistId = "therapist_id"
        case sessionId = "session_id"
        case type
    }
}

/// Ongoing diary assignment. `session_id` and completion timestamps stay NULL
/// until the therapist cancels (`cancelled_at` only).
private struct NewOngoingPatientAssignment: Encodable {
    let patientId: UUID
    let therapistId: UUID
    let type: PatientAssignmentType

    enum CodingKeys: String, CodingKey {
        case patientId = "patient_id"
        case therapistId = "therapist_id"
        case sessionId = "session_id"
        case type
        case completedAt = "completed_at"
        case cancelledAt = "cancelled_at"
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(patientId, forKey: .patientId)
        try container.encode(therapistId, forKey: .therapistId)
        try container.encodeNil(forKey: .sessionId)
        try container.encode(type, forKey: .type)
        try container.encodeNil(forKey: .completedAt)
        try container.encodeNil(forKey: .cancelledAt)
    }
}

private struct CancelPatientAssignment: Encodable {
    let cancelledAt: String

    enum CodingKeys: String, CodingKey {
        case cancelledAt = "cancelled_at"
    }
}

private struct IsPatientConnectedParams: Encodable {
    let pPatientId: UUID

    enum CodingKeys: String, CodingKey {
        case pPatientId = "p_patient_id"
    }
}

/// Therapist-side patient assignments. Connection is decided only by the
/// `is_patient_connected` RPC — never by reading `patient_access`.
@MainActor
final class PatientAssignmentService {
    private let client: SupabaseClient

    init(client: SupabaseClient) {
        self.client = client
    }

    /// Active Patient Mode connection for this patient. Boolean only.
    func isPatientConnected(patientId: UUID) async throws -> Bool {
        try ensureConfigured()
        do {
            let connected: Bool = try await client.rpc(
                "is_patient_connected",
                params: IsPatientConnectedParams(pPatientId: patientId)
            )
            .execute()
            .value
            return connected
        } catch {
            AppLog.store.error(
                "Patient connection check failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    /// Open questionnaire assignment for this session, if one exists.
    func openQuestionnaireAssignment(sessionId: UUID) async throws -> PatientAssignment? {
        try ensureConfigured()
        do {
            let rows: [PatientAssignment] = try await client.from("patient_assignments")
                .select(
                    "id, patient_id, therapist_id, session_id, type, created_at, completed_at, cancelled_at"
                )
                .eq("session_id", value: sessionId)
                .eq("type", value: PatientAssignmentType.questionnaire.rawValue)
                .is("completed_at", value: nil)
                .is("cancelled_at", value: nil)
                .limit(1)
                .execute()
                .value
            return rows.first
        } catch {
            AppLog.store.error(
                "Questionnaire assignment fetch failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    /// Creates a one-time questionnaire assignment after a connection and
    /// duplicate check. Does not write CombinedMood / questionnaire answers.
    func sendQuestionnaireAssignment(
        patientId: UUID,
        sessionId: UUID
    ) async throws -> PatientAssignment {
        try ensureConfigured()
        let connected = try await isPatientConnected(patientId: patientId)
        guard connected else { throw PatientAssignmentError.patientNotConnected }

        if let existing = try await openQuestionnaireAssignment(sessionId: sessionId) {
            return existing
        }

        let therapistId = try await requireTherapistId()
        do {
            let created: PatientAssignment = try await client.from("patient_assignments")
                .insert(
                    NewPatientAssignment(
                        patientId: patientId,
                        therapistId: therapistId,
                        sessionId: sessionId,
                        type: .questionnaire
                    )
                )
                .select(
                    "id, patient_id, therapist_id, session_id, type, created_at, completed_at, cancelled_at"
                )
                .single()
                .execute()
                .value
            AppLog.store.info("Questionnaire assignment created")
            return created
        } catch {
            AppLog.store.error(
                "Questionnaire assignment create failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    /// Active ongoing diary assignment (`cancelled_at` is NULL). `completed_at`
    /// is not part of diary active-state.
    func activeOngoingAssignment(
        patientId: UUID,
        type: PatientAssignmentType
    ) async throws -> PatientAssignment? {
        try ensureConfigured()
        try Self.requireOngoingType(type)
        do {
            let rows: [PatientAssignment] = try await client.from("patient_assignments")
                .select(
                    "id, patient_id, therapist_id, session_id, type, created_at, completed_at, cancelled_at"
                )
                .eq("patient_id", value: patientId)
                .eq("type", value: type.rawValue)
                .is("cancelled_at", value: nil)
                .limit(1)
                .execute()
                .value
            return rows.first
        } catch {
            AppLog.store.error(
                "Ongoing assignment fetch failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    /// Inserts a new ongoing assignment. Never reopens a cancelled row.
    func activateOngoingAssignment(
        patientId: UUID,
        type: PatientAssignmentType
    ) async throws -> PatientAssignment {
        try ensureConfigured()
        try Self.requireOngoingType(type)
        let connected = try await isPatientConnected(patientId: patientId)
        guard connected else { throw PatientAssignmentError.patientNotConnected }

        if let existing = try await activeOngoingAssignment(patientId: patientId, type: type) {
            return existing
        }

        let therapistId = try await requireTherapistId()
        do {
            let created: PatientAssignment = try await client.from("patient_assignments")
                .insert(
                    NewOngoingPatientAssignment(
                        patientId: patientId,
                        therapistId: therapistId,
                        type: type
                    )
                )
                .select(
                    "id, patient_id, therapist_id, session_id, type, created_at, completed_at, cancelled_at"
                )
                .single()
                .execute()
                .value
            AppLog.store.info("Ongoing assignment created")
            return created
        } catch {
            if Self.isUniqueViolation(error),
               let existing = try? await activeOngoingAssignment(patientId: patientId, type: type) {
                AppLog.store.info("Ongoing assignment already active")
                return existing
            }
            AppLog.store.error(
                "Ongoing assignment create failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    /// Stops an ongoing assignment by setting `cancelled_at` only.
    func cancelOngoingAssignment(id: UUID) async throws {
        try ensureConfigured()
        do {
            let updated: [PatientAssignment] = try await client.from("patient_assignments")
                .update(
                    CancelPatientAssignment(cancelledAt: Self.timestampString(from: Date()))
                )
                .eq("id", value: id)
                .is("cancelled_at", value: nil)
                .select(
                    "id, patient_id, therapist_id, session_id, type, created_at, completed_at, cancelled_at"
                )
                .execute()
                .value
            guard !updated.isEmpty else {
                throw PatientAssignmentError.invalidIdentifier
            }
            AppLog.store.info("Ongoing assignment cancelled")
        } catch {
            AppLog.store.error(
                "Ongoing assignment cancel failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    /// Patient Mode list. RLS limits rows to this auth user's Patient;
    /// `patientId` is an optional extra filter, not authorization.
    func patientAssignments(patientId: UUID? = nil) async throws -> [PatientAssignment] {
        try ensureConfigured()
        do {
            let filter = client.from("patient_assignments")
                .select(
                    "id, patient_id, session_id, type, created_at, completed_at, cancelled_at"
                )
                .is("cancelled_at", value: nil)
            let rows: [PatientAssignment]
            if let patientId {
                rows = try await filter
                    .eq("patient_id", value: patientId)
                    .order("created_at", ascending: false)
                    .execute()
                    .value
            } else {
                rows = try await filter
                    .order("created_at", ascending: false)
                    .execute()
                    .value
            }
            AppLog.store.info("Patient assignments loaded")
            return rows
        } catch {
            AppLog.store.error(
                "Patient assignments fetch failed: \(error.localizedDescription, privacy: .public)"
            )
            throw error
        }
    }

    /// Submits GAD-7 / PHQ-9 answers for an open assignment. CombinedMood is
    /// created only by the Edge Function — never from this client.
    func submitPatientQuestionnaire(
        assignmentId: UUID,
        gad7Answers: [Int],
        phq9Answers: [Int],
        interferenceLevel: Int
    ) async throws {
        try ensureConfigured()
        try Self.validatePatientAnswers(
            gad7Answers: gad7Answers,
            phq9Answers: phq9Answers,
            interferenceLevel: interferenceLevel
        )
        do {
            let response: SubmitPatientQuestionnaireResponse = try await client.functions.invoke(
                "submit-patient-questionnaire",
                options: FunctionInvokeOptions(
                    body: SubmitPatientQuestionnaireRequest(
                        assignmentId: assignmentId,
                        gad7Answers: gad7Answers,
                        phq9Answers: phq9Answers,
                        interferenceLevel: interferenceLevel
                    )
                )
            )
            guard response.success != false else {
                throw PatientQuestionnaireSubmitError.failed
            }
            AppLog.store.info("Patient questionnaire submitted")
        } catch let error as PatientQuestionnaireSubmitError {
            throw error
        } catch let FunctionsError.httpError(code, data) {
            throw Self.submitError(from: data, statusCode: code)
        } catch {
            AppLog.store.error(
                "Patient questionnaire submit failed: \(error.localizedDescription, privacy: .public)"
            )
            throw PatientQuestionnaireSubmitError.failed
        }
    }

    private static func validatePatientAnswers(
        gad7Answers: [Int],
        phq9Answers: [Int],
        interferenceLevel: Int
    ) throws {
        let valid = CombinedMoodQuestionnaire.answerValues
        guard gad7Answers.count == L10n.gad7Questions.count,
              phq9Answers.count == L10n.phq9Questions.count,
              gad7Answers.allSatisfy({ valid.contains($0) }),
              phq9Answers.allSatisfy({ valid.contains($0) }),
              valid.contains(interferenceLevel)
        else {
            throw PatientQuestionnaireSubmitError.invalidAnswers
        }
    }

    private static func submitError(from data: Data, statusCode: Int) -> PatientQuestionnaireSubmitError {
        let code = edgeErrorCode(from: data)
        switch code {
        case "assignment_already_completed", "already_completed":
            return .alreadyCompleted
        case "assignment_cancelled", "cancelled":
            return .cancelled
        case "access_denied", "forbidden", "unauthorized":
            return .accessDenied
        default:
            if statusCode == 401 || statusCode == 403 {
                return .accessDenied
            }
            AppLog.store.error("Patient questionnaire submit failed")
            return .failed
        }
    }

    private static func edgeErrorCode(from data: Data) -> String {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return ""
        }
        if let error = json["error"] as? String { return error }
        if let code = json["code"] as? String { return code }
        if let nested = json["error"] as? [String: Any],
           let code = nested["code"] as? String {
            return code
        }
        if let status = json["status"] as? String { return status }
        return ""
    }

    private func requireTherapistId() async throws -> UUID {
        do {
            return try await client.auth.session.user.id
        } catch {
            throw PatientAssignmentError.notSignedIn
        }
    }

    private func ensureConfigured() throws {
        guard SupabaseConfig.isConfigured else { throw PatientAssignmentError.notConfigured }
    }

    private static func requireOngoingType(_ type: PatientAssignmentType) throws {
        switch type {
        case .diaryOne, .diaryTwo:
            return
        case .questionnaire:
            throw PatientAssignmentError.invalidIdentifier
        }
    }

    private static func isUniqueViolation(_ error: Error) -> Bool {
        if let postgrest = error as? PostgrestError, postgrest.code == "23505" {
            return true
        }
        let nsError = error as NSError
        if nsError.code == 23505 { return true }
        if let nested = nsError.userInfo[NSUnderlyingErrorKey] as? Error {
            return isUniqueViolation(nested)
        }
        return false
    }

    private static func timestampString(from date: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: date)
    }
}

extension DatabaseID {
    var uuidValue: UUID? {
        UUID(uuidString: queryValue)
    }
}

File: android/app/src/main/java/com/cbtipul/app/data/PatientAssignment.kt
package com.cbtipul.app.data

import com.cbtipul.app.model.DatabaseId
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.functions.functions
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.Headers
import io.ktor.http.HttpHeaders
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import java.text.SimpleDateFormat
import java.time.Instant
import java.time.OffsetDateTime
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID

enum class PatientAssignmentType(val raw: String) {
    Questionnaire("questionnaire"),
    DiaryOne("diary_one"),
    DiaryTwo("diary_two"),
    ;

    companion object {
        fun fromRaw(value: String): PatientAssignmentType? = entries.find { it.raw == value }
    }
}

data class PatientAssignment(
    val id: String,
    val patientId: String,
    val therapistId: String?,
    val sessionId: String?,
    val typeValue: String,
    val createdAt: Date,
    val completedAt: Date?,
    val cancelledAt: Date?,
) {
    val type: PatientAssignmentType? get() = PatientAssignmentType.fromRaw(typeValue)
    val isOpen: Boolean get() = completedAt == null && cancelledAt == null
}

@Serializable
private data class PatientAssignmentRow(
    val id: String,
    @SerialName("patient_id") val patientId: String,
    @SerialName("therapist_id") val therapistId: String? = null,
    @SerialName("session_id") val sessionId: String? = null,
    val type: String,
    @SerialName("created_at") val createdAt: String,
    @SerialName("completed_at") val completedAt: String? = null,
    @SerialName("cancelled_at") val cancelledAt: String? = null,
)

@Serializable
internal data class RequestPatientQuestionnaireRequest(
    val patientId: String,
    val sessionId: String,
)

@Serializable
private data class SubmitPatientQuestionnaireRequest(
    val assignmentId: String,
    val gad7Answers: List<Int>,
    val phq9Answers: List<Int>,
    val interferenceLevel: Int,
)

@Serializable
private data class SubmitPatientQuestionnaireResponse(
    val success: Boolean? = null,
    val combinedMoodId: Int? = null,
    val sessionId: String? = null,
)

sealed class PatientQuestionnaireSubmitError : Exception() {
    data object AlreadyCompleted : PatientQuestionnaireSubmitError()
    data object Cancelled : PatientQuestionnaireSubmitError()
    data object AccessDenied : PatientQuestionnaireSubmitError()
    data object InvalidAnswers : PatientQuestionnaireSubmitError()
    data object Failed : PatientQuestionnaireSubmitError()
}

sealed class PatientAssignmentException : Exception() {
    data object NotConfigured : PatientAssignmentException()
    data object NotSignedIn : PatientAssignmentException()
    data object InvalidIdentifier : PatientAssignmentException()
    data object PatientNotConnected : PatientAssignmentException()
}

class PatientAssignmentRepository(private val client: SupabaseClient) {
    suspend fun isPatientConnected(patientId: String): Boolean {
        ensureConfigured()
        val result = client.postgrest.rpc(
            "is_patient_connected",
            buildJsonObject { put("p_patient_id", patientId) },
        )
        return Json.parseToJsonElement(result.data).jsonPrimitive.boolean
    }

    suspend fun openQuestionnaireAssignment(sessionId: String): PatientAssignment? {
        ensureConfigured()
        val rows = client.from("patient_assignments")
            .select(assignmentColumns) {
                filter {
                    eq("session_id", sessionId)
                    eq("type", PatientAssignmentType.Questionnaire.raw)
                    exact("completed_at", null)
                    exact("cancelled_at", null)
                }
                limit(1)
            }
            .decodeList<PatientAssignmentRow>()
        return rows.firstOrNull()?.toDomain()
    }

    suspend fun sendQuestionnaireAssignment(patientId: String, sessionId: String): PatientAssignment {
        ensureConfigured()
        return try {
            val http = client.functions.invoke(
                function = "request-patient-questionnaire",
                body = RequestPatientQuestionnaireRequest(
                    patientId = patientId,
                    sessionId = sessionId,
                ),
                headers = Headers.build {
                    append(HttpHeaders.ContentType, ContentType.Application.Json.toString())
                },
            )
            assignmentFromEdgeJson(http.bodyAsText())
        } catch (error: PatientAssignmentException) {
            throw error
        } catch (error: Exception) {
            throw mapRequestQuestionnaireError(error)
        }
    }

    suspend fun activeOngoingAssignment(patientId: String, type: PatientAssignmentType): PatientAssignment? {
        ensureConfigured()
        requireOngoing(type)
        val rows = client.from("patient_assignments")
            .select(assignmentColumns) {
                filter {
                    eq("patient_id", patientId)
                    eq("type", type.raw)
                    exact("cancelled_at", null)
                }
                limit(1)
            }
            .decodeList<PatientAssignmentRow>()
        return rows.firstOrNull()?.toDomain()
    }

    suspend fun activateOngoingAssignment(patientId: String, type: PatientAssignmentType): PatientAssignment {
        ensureConfigured()
        requireOngoing(type)
        if (!isPatientConnected(patientId)) throw PatientAssignmentException.PatientNotConnected
        activeOngoingAssignment(patientId, type)?.let { return it }
        val therapistId = requireTherapistId()
        val body = buildJsonObject {
            put("patient_id", patientId)
            put("therapist_id", therapistId)
            put("session_id", JsonNull)
            put("type", type.raw)
            put("completed_at", JsonNull)
            put("cancelled_at", JsonNull)
        }
        return try {
            client.from("patient_assignments")
                .insert(body) { select(assignmentColumns) }
                .decodeSingle<PatientAssignmentRow>()
                .toDomain()
        } catch (error: Exception) {
            if (isUniqueViolation(error)) {
                activeOngoingAssignment(patientId, type)?.let { return it }
            }
            throw error
        }
    }

    suspend fun cancelOngoingAssignment(id: String) {
        ensureConfigured()
        val body = buildJsonObject { put("cancelled_at", timestampNow()) }
        val updated = client.from("patient_assignments")
            .update(body) {
                filter {
                    eq("id", id)
                    exact("cancelled_at", null)
                }
                select(assignmentColumns)
            }
            .decodeList<PatientAssignmentRow>()
        if (updated.isEmpty()) throw PatientAssignmentException.InvalidIdentifier
    }

    suspend fun patientAssignments(patientId: String?): List<PatientAssignment> {
        ensureConfigured()
        val rows = client.from("patient_assignments")
            .select(
                Columns.raw("id, patient_id, session_id, type, created_at, completed_at, cancelled_at"),
            ) {
                filter {
                    exact("cancelled_at", null)
                    if (patientId != null) eq("patient_id", patientId)
                }
                order("created_at", Order.DESCENDING)
            }
            .decodeList<PatientAssignmentRow>()
        return rows.map { it.toDomain() }
    }

    suspend fun submitPatientQuestionnaire(
        assignmentId: String,
        gad7Answers: List<Int>,
        phq9Answers: List<Int>,
        interferenceLevel: Int,
    ) {
        ensureConfigured()
        val valid = (0..3).toSet()
        if (gad7Answers.size != 7 || phq9Answers.size != 9 ||
            gad7Answers.any { it !in valid } || phq9Answers.any { it !in valid } ||
            interferenceLevel !in valid
        ) {
            throw PatientQuestionnaireSubmitError.InvalidAnswers
        }
        try {
            val http = client.functions.invoke(
                function = "submit-patient-questionnaire",
                body = SubmitPatientQuestionnaireRequest(
                    assignmentId = assignmentId,
                    gad7Answers = gad7Answers,
                    phq9Answers = phq9Answers,
                    interferenceLevel = interferenceLevel,
                ),
                headers = Headers.build {
                    append(HttpHeaders.ContentType, ContentType.Application.Json.toString())
                },
            )
            val response = EdgePayload.json.decodeFromString(
                SubmitPatientQuestionnaireResponse.serializer(),
                http.bodyAsText(),
            )
            if (response.success == false) throw PatientQuestionnaireSubmitError.Failed
        } catch (error: PatientQuestionnaireSubmitError) {
            throw error
        } catch (error: Exception) {
            throw mapQuestionnaireError(error)
        }
    }

    private suspend fun mapQuestionnaireError(error: Exception): PatientQuestionnaireSubmitError {
        val body = EdgePayload.responseBody(error)
        val (code, _) = EdgePayload.codeAndMessage(body)
        val status = EdgePayload.httpStatus(error)
        return when (code) {
            "assignment_already_completed", "already_completed" -> PatientQuestionnaireSubmitError.AlreadyCompleted
            "assignment_cancelled", "cancelled" -> PatientQuestionnaireSubmitError.Cancelled
            "access_denied", "forbidden", "unauthorized" -> PatientQuestionnaireSubmitError.AccessDenied
            else -> if (status == 401 || status == 403) {
                PatientQuestionnaireSubmitError.AccessDenied
            } else {
                PatientQuestionnaireSubmitError.Failed
            }
        }
    }

    private suspend fun mapRequestQuestionnaireError(error: Exception): PatientAssignmentException {
        val body = EdgePayload.responseBody(error)
        val (code, _) = EdgePayload.codeAndMessage(body)
        val status = EdgePayload.httpStatus(error)
        return mapRequestQuestionnaireCode(code, status)
    }

    private suspend fun requireTherapistId(): String {
        return client.auth.currentSessionOrNull()?.user?.id
            ?: throw PatientAssignmentException.NotSignedIn
    }

    private fun ensureConfigured() {
        if (!SupabaseConfig.isConfigured) throw PatientAssignmentException.NotConfigured
    }

    private fun requireOngoing(type: PatientAssignmentType) {
        if (type == PatientAssignmentType.Questionnaire) {
            throw PatientAssignmentException.InvalidIdentifier
        }
    }

    private fun isUniqueViolation(error: Throwable): Boolean {
        return error.message?.contains("23505") == true
    }

    private fun timestampNow(): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale.US)
        formatter.timeZone = TimeZone.getTimeZone("UTC")
        return formatter.format(Date())
    }

    companion object {
        private val assignmentColumns = Columns.raw(
            "id, patient_id, therapist_id, session_id, type, created_at, completed_at, cancelled_at",
        )

        fun uuidOrNull(id: DatabaseId): String? =
            runCatching { UUID.fromString(id.queryValue).toString() }.getOrNull()

        internal fun encodeQuestionnaireRequest(patientId: String, sessionId: String): String =
            EdgePayload.json.encodeToString(
                RequestPatientQuestionnaireRequest.serializer(),
                RequestPatientQuestionnaireRequest(patientId = patientId, sessionId = sessionId),
            )

        internal fun assignmentFromEdgeJson(json: String): PatientAssignment =
            EdgePayload.json.decodeFromString(PatientAssignmentRow.serializer(), json).toDomain()

        internal fun mapRequestQuestionnaireCode(code: String, status: Int?): PatientAssignmentException {
            return when (code) {
                "patient_not_connected" -> PatientAssignmentException.PatientNotConnected
                "unauthorized" -> PatientAssignmentException.NotSignedIn
                else -> if (status == 401 || status == 403) {
                    PatientAssignmentException.NotSignedIn
                } else {
                    PatientAssignmentException.InvalidIdentifier
                }
            }
        }

        internal fun parseAssignmentTimestamp(raw: String): Date {
            val trimmed = raw.trim()
            require(trimmed.isNotEmpty())
            runCatching { Date.from(OffsetDateTime.parse(trimmed).toInstant()) }.getOrNull()?.let { return it }
            runCatching { Date.from(Instant.parse(trimmed)) }.getOrNull()?.let { return it }
            listOf(
                SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale.US),
                SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX", Locale.US),
                SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX", Locale.US),
                SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssX", Locale.US),
            ).forEach { formatter ->
                formatter.timeZone = TimeZone.getTimeZone("UTC")
                formatter.parse(trimmed)?.let { return it }
            }
            throw IllegalArgumentException("invalid_timestamp")
        }

        private fun parseIso(raw: String): Date = parseAssignmentTimestamp(raw)

        private fun PatientAssignmentRow.toDomain() = PatientAssignment(
            id = id,
            patientId = patientId,
            therapistId = therapistId,
            sessionId = sessionId,
            typeValue = type,
            createdAt = parseIso(createdAt),
            completedAt = completedAt?.let(::parseIso),
            cancelledAt = cancelledAt?.let(::parseIso),
        )
    }
}
